File: backend/internal/handlers/identities.go
package handlers

import (
	"net/http"
	"strconv"

	"github.com/benoit-sauvere/kratos-admin-ui/backend/internal/kratos"
	"github.com/gin-gonic/gin"
	ory "github.com/ory/kratos-client-go"
)

// IdentitiesHandler handles identity-related requests
type IdentitiesHandler struct {
	client *kratos.Client
}

// NewIdentitiesHandler creates a new identities handler
func NewIdentitiesHandler(client *kratos.Client) *IdentitiesHandler {
	return &IdentitiesHandler{client: client}
}

// List returns a paginated list of identities
func (h *IdentitiesHandler) List(c *gin.Context) {
	page, _ := strconv.ParseInt(c.DefaultQuery("page", "1"), 10, 64)
	perPage, _ := strconv.ParseInt(c.DefaultQuery("per_page", "20"), 10, 64)

	result, err := h.client.ListIdentities(c.Request.Context(), page, perPage)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch identities", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"data":     result.Identities,
		"page":     page,
		"per_page": perPage,
		"total":    result.Total,
	})
}

// Get returns a single identity by ID
func (h *IdentitiesHandler) Get(c *gin.Context) {
	id := c.Param("id")

	identity, err := h.client.GetIdentity(c.Request.Context(), id)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Identity not found", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, identity)
}

// CreateIdentityRequest represents the request body for creating an identity
type CreateIdentityRequest struct {
	SchemaID string                 `json:"schema_id" binding:"required"`
	Traits   map[string]interface{} `json:"traits" binding:"required"`
	State    string                 `json:"state,omitempty"`
}

// Create creates a new identity
func (h *IdentitiesHandler) Create(c *gin.Context) {
	var req CreateIdentityRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body", "details": err.Error()})
		return
	}

	body := ory.CreateIdentityBody{
		SchemaId: req.SchemaID,
		Traits:   req.Traits,
	}

	if req.State != "" {
		state := ory.IdentityState(req.State)
		body.State = &state
	}

	identity, err := h.client.CreateIdentity(c.Request.Context(), body)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create identity", "details": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, identity)
}

// UpdateIdentityRequest represents the request body for updating an identity
type UpdateIdentityRequest struct {
	SchemaID string                 `json:"schema_id" binding:"required"`
	Traits   map[string]interface{} `json:"traits" binding:"required"`
	State    string                 `json:"state,omitempty"`
}

// Update updates an existing identity
func (h *IdentitiesHandler) Update(c *gin.Context) {
	id := c.Param("id")

	var req UpdateIdentityRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body", "details": err.Error()})
		return
	}

	body := ory.UpdateIdentityBody{
		SchemaId: req.SchemaID,
		Traits:   req.Traits,
	}

	if req.State != "" {
		state := ory.IdentityState(req.State)
		body.State = state
	}

	identity, err := h.client.UpdateIdentity(c.Request.Context(), id, body)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update identity", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, identity)
}

// Delete deletes an identity
func (h *IdentitiesHandler) Delete(c *gin.Context) {
	id := c.Param("id")

	if err := h.client.DeleteIdentity(c.Request.Context(), id); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete identity", "details": err.Error()})
		return
	}

	c.JSON(http.StatusNoContent, nil)
}

// GetSessions returns sessions for an identity
func (h *IdentitiesHandler) GetSessions(c *gin.Context) {
	id := c.Param("id")

	sessions, err := h.client.GetIdentitySessions(c.Request.Context(), id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch sessions", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"data": sessions})
}

// ResetPasswordRequest represents the request body for resetting a password
type ResetPasswordRequest struct {
	Password string `json:"password" binding:"required,min=8"`
}

// ResetPassword resets the password for an identity
func (h *IdentitiesHandler) ResetPassword(c *gin.Context) {
	id := c.Param("id")

	var req ResetPasswordRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body", "details": err.Error()})
		return
	}

	if err := h.client.ResetPassword(c.Request.Context(), id, req.Password); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to reset password", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Password reset successfully"})
}

// DeleteCredential deletes a specific credential type for an identity
func (h *IdentitiesHandler) DeleteCredential(c *gin.Context) {
	id := c.Param("id")
	credType := c.Param("type")

	// Validate credential type
	validTypes := map[string]bool{
		"totp":          true,
		"webauthn":      true,
		"lookup_secret": true,
	}

	if !validTypes[credType] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid credential type", "details": "Supported types: totp, webauthn, lookup_secret"})
		return
	}

	if err := h.client.DeleteCredential(c.Request.Context(), id, credType); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete credential", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Credential deleted successfully"})
}

// GetWithCredentials returns a single identity by ID including credentials metadata
func (h *IdentitiesHandler) GetWithCredentials(c *gin.Context) {
	id := c.Param("id")

	identity, err := h.client.GetIdentityWithCredentials(c.Request.Context(), id)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Identity not found", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, identity)
}

File: frontend/src/stores/identities.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '@/api/client'
import type { Identity, Session } from '@/types'

export const useIdentitiesStore = defineStore('identities', () => {
  const identities = ref<Identity[]>([])
  const currentIdentity = ref<Identity | null>(null)
  const currentIdentitySessions = ref<Session[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)
  const page = ref(1)
  const perPage = ref(20)
  const total = ref(0)

  async function fetchIdentities(pageNum = 1, itemsPerPage = 20) {
    loading.value = true
    error.value = null
    try {
      const response = await api.getIdentities(pageNum, itemsPerPage)
      identities.value = response.data
      page.value = response.page
      perPage.value = response.per_page
      total.value = response.total || response.data.length
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to fetch identities'
    } finally {
      loading.value = false
    }
  }

  async function fetchIdentity(id: string) {
    loading.value = true
    error.value = null
    try {
      currentIdentity.value = await api.getIdentity(id)
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to fetch identity'
    } finally {
      loading.value = false
    }
  }

  async function fetchIdentitySessions(id: string) {
    loading.value = true
    error.value = null
    try {
      const response = await api.getIdentitySessions(id)
      currentIdentitySessions.value = response.data
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to fetch sessions'
    } finally {
      loading.value = false
    }
  }

  async function createIdentity(data: { schema_id: string; traits: Record<string, unknown>; state?: string }) {
    loading.value = true
    error.value = null
    try {
      const identity = await api.createIdentity(data)
      identities.value.unshift(identity)
      return identity
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to create identity'
      throw e
    } finally {
      loading.value = false
    }
  }

  async function updateIdentity(id: string, data: { schema_id: string; traits: Record<string, unknown>; state?: string }) {
    loading.value = true
    error.value = null
    try {
      const identity = await api.updateIdentity(id, data)
      const index = identities.value.findIndex(i => i.id === id)
      if (index !== -1) {
        identities.value[index] = identity
      }
      if (currentIdentity.value?.id === id) {
        currentIdentity.value = identity
      }
      return identity
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to update identity'
      throw e
    } finally {
      loading.value = false
    }
  }

  async function deleteIdentity(id: string) {
    loading.value = true
    error.value = null
    try {
      await api.deleteIdentity(id)
      identities.value = identities.value.filter(i => i.id !== id)
      if (currentIdentity.value?.id === id) {
        currentIdentity.value = null
      }
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to delete identity'
      throw e
    } finally {
      loading.value = false
    }
  }

  async function fetchIdentityWithCredentials(id: string) {
    loading.value = true
    error.value = null
    try {
      currentIdentity.value = await api.getIdentityWithCredentials(id)
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to fetch identity'
    } finally {
      loading.value = false
    }
  }

  async function resetPassword(id: string, password: string) {
    loading.value = true
    error.value = null
    try {
      await api.resetPassword(id, password)
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to reset password'
      throw e
    } finally {
      loading.value = false
    }
  }

  async function deleteCredential(id: string, credentialType: string) {
    loading.value = true
    error.value = null
    try {
      await api.deleteCredential(id, credentialType)
      // Refresh the identity to get updated credentials
      await fetchIdentityWithCredentials(id)
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to delete credential'
      throw e
    } finally {
      loading.value = false
    }
  }

  return {
    identities,
    currentIdentity,
    currentIdentitySessions,
    loading,
    error,
    page,
    perPage,
    total,
    fetchIdentities,
    fetchIdentity,
    fetchIdentitySessions,
    fetchIdentityWithCredentials,
    createIdentity,
    updateIdentity,
    deleteIdentity,
    resetPassword,
    deleteCredential
  }
})




